src/extract/faster_bottom_up.rs

```diff
@@ -0,0 +1,117 @@
+use super::*;
+
+/// A faster bottom up extractor inspired by the faster-greedy-dag extractor.
+pub struct BottomUpExtractor;
+
+impl Extractor for BottomUpExtractor {
+    fn extract(&self, egraph: &EGraph, _roots: &[ClassId]) -> ExtractionResult {
+        // 1. build map from class to parent nodes
+        let mut parents = IndexMap::<ClassId, Vec<NodeId>>::default();
+        let n2c = |nid: &NodeId| egraph.nid_to_cid(nid);
+
+        for class in egraph.classes().values() {
+            parents.insert(class.id.clone(), Vec::new());
+        }
+        for class in egraph.classes().values() {
+            for node in &class.nodes {
+                for c in &egraph[node].children {
+                    parents[n2c(c)].push(node.clone());
+                }
+            }
+        }
+
+        // 2. start analysis from leaves
```

Comment:
optional: Same question: can this be in the same loop?

Reply:
Moved it to the other loop